Stripe/PublicHeaders/STPEphemeralKeyProvider.h

```diff
@@ -40,7 +39,44 @@ NS_ASSUME_NONNULL_BEGIN
  or `completion(nil, error)` if an error is returned.
  */
 - (void)createCustomerKeyWithAPIVersion:(NSString *)apiVersion completion:(STPJSONResponseCompletionBlock)completion;
+@end
 
+/**
+ You should make your application's API client conform to this interface.
+ It provides a way for Stripe utility classes to request a new ephemeral key from
+ your backend, which it will use to retrieve and update Stripe API objects.
+ */
+@protocol STPIssuingCardEphemeralKeyProvider <NSObject>
```

Comment:
I'm not sure these need to be two separate protocols. The documentation says that the API client should conform to both of these, which means checking for `keyProvider` protocol conformance will always succeed for both. 

It's not totally clear to me how or why we intend to differentiate between creating customer and card ephemeral keys? Is the intention for the integrator to create two instances of `STPEphemeralKeyManager` and call the right one as needed?

Reply:
@jack-stripe ?

Reply:
That's the way I imagined it, yes. You might imagine one API that calls `POST https://myapi.com/users/:id/ephemeral_keys` and one that calls `POST https://myapi.com/cards/:id/ephemeral_keys`. Then you might have one API client client-side that calls the first, and another that calls the second.

The other alternative, as you suggest, is to make a single API - `POST https://myapi.com/ephemeral_keys` (with `user` or `card` parameters) and a single API client to talk to both. The issue is that now the method signature needs to change - we need to tell the app what they should be making an ephemeral key for exactly. So the protocol method would have to be something generic like `createKey:forAssociatedStripeObject:` which just feels like turning a thing to figure out at compile-time into a thing to figure out at runtime.

You raise a good point, though, that this breaks if someone's API client conforms to both protocols. Sadly, I think we'll need to add a property to `STPEphemeralKeyManager` to indicate what kind of key it is intended to make (not the worst thing, as it's a private class). Thoughts @csabol-stripe ?

Reply:
hmm, what about adding a new method to `STPEphemeralKeyManager` so we have `getCustomerKey:` and `getCardKey:` (or similar)? That way we only have one key manager but the requests are still "typed" at compile time. 

Similarly, what are your thoughts about keeping `STPEphemeralKeyProvider` and just adding `createIssuingCardKeyWithAPIVersion:completion:` to the existing protocol (and marking as optional if we think that would be useful?)

Reply:
I think I'd prefer to keep a property on `STPEphemeralKeyManager`, because I think it'd be a programmer error to call `getCustomerKey:` and then later `getCardKey:` on the same instance. Nice thing is, assuming that property is set in the initializer, we could assert that the given `keyProvider` conforms to the necessary protocol.

I don't feel as strongly about whether or not to split `STPEphemeralKeyProvider` into 2 protocols, but realistically users are going to be implementing only one of `createCustomerKeyWithAPIVersion:completion` or createIssuingCardKeyWithAPIVersion:completion`, so both would have to be optional. In that case, I think it's a little nicer to just have two protocols so that you can enforce at compile-time (instead of runtime) that the necessary methods are implemented.

Reply:
Thoughts @csabol-stripe? Can also discuss in slack on monday : )

Reply:
closing the loop: @jack-stripe and I met to discuss this today and we are happy moving forward with the patch as is :)